File: backend/src/main/java/com/github/aadvorak/artilleryonline/battle/calculator/VehicleAccelerationCalculator.java
package com.github.aadvorak.artilleryonline.battle.calculator;

import com.github.aadvorak.artilleryonline.battle.calculations.ForceCalculator;
import com.github.aadvorak.artilleryonline.battle.calculations.VehicleCalculations;
import com.github.aadvorak.artilleryonline.battle.calculator.vehicle.*;
import com.github.aadvorak.artilleryonline.battle.common.BodyAcceleration;
import com.github.aadvorak.artilleryonline.battle.config.VehicleConfig;
import com.github.aadvorak.artilleryonline.battle.model.BattleModel;
import com.github.aadvorak.artilleryonline.battle.model.VehicleModel;
import com.github.aadvorak.artilleryonline.battle.precalc.VehiclePreCalc;
import com.github.aadvorak.artilleryonline.battle.specs.VehicleSpecs;
import com.github.aadvorak.artilleryonline.battle.state.VehicleState;

import java.util.List;

public class VehicleAccelerationCalculator {

    private static final List<
            ForceCalculator<VehicleSpecs, VehiclePreCalc, VehicleConfig, VehicleState, VehicleModel, VehicleCalculations>
            > forceCalculators = List.of(
            new GravityForceCalculator(),
            new GroundFrictionForceCalculator(),
            new GroundReactionForceCalculator(),
            new JetForceCalculator(),
            new EngineForceCalculator()
    );

    private static final BodyAccelerationCalculator<
            VehicleSpecs, VehiclePreCalc, VehicleConfig, VehicleState, VehicleModel, VehicleCalculations
            > calculator = new BodyAccelerationCalculator<>(forceCalculators);

    public static BodyAcceleration getVehicleAcceleration(VehicleCalculations vehicle, BattleModel battleModel) {
        vehicle.calculateAllGroundContacts(battleModel.getRoom());
        return calculator.calculate(vehicle, battleModel);
    }
}

File: frontend/playground/battle/calculator/vehicle-acceleration-calculator.ts
import type {BodyAcceleration} from '@/playground/data/common'
import type {BattleModel} from '@/playground/data/model'
import type {VehicleCalculations} from '@/playground/data/calculations'
import {VehicleUtils} from '@/playground/utils/vehicle-utils'
import {BodyAccelerationCalculator} from "~/playground/battle/calculator/body-acceleration-calculator";
import {GroundFrictionForceCalculator} from "~/playground/battle/calculator/vehicle/ground-friction-force-calculator";
import {GroundReactionForceCalculator} from "~/playground/battle/calculator/vehicle/ground-reaction-force-calculator";
import {GravityForceCalculator} from "~/playground/battle/calculator/vehicle/gravity-force-calculator";
import {JetForceCalculator} from "~/playground/battle/calculator/vehicle/jet-force-calculator";
import {EngineForceCalculator} from "~/playground/battle/calculator/vehicle/engine-force-calculator";

export const VehicleAccelerationCalculator = {
  calculator: new BodyAccelerationCalculator<VehicleCalculations>(
      [
        new GravityForceCalculator(),
        new GroundFrictionForceCalculator(),
        new GroundReactionForceCalculator(),
        new JetForceCalculator(),
        new EngineForceCalculator()
      ]
  ),

  getVehicleAcceleration(vehicle: VehicleCalculations, battleModel: BattleModel): BodyAcceleration {
    VehicleUtils.calculateAllGroundContacts(vehicle, battleModel.room)
    return this.calculator.calculate(vehicle, battleModel)
  },
}
